File: backend/internal/cli/crew_run.go
package cli

import (
	"errors"
	"fmt"
	"io"
	"net/url"
	"os"
	"strings"
	"time"

	"github.com/spf13/cobra"
)

// crewRunView mirrors the daemon's domain.CrewRun on the wire. Hand-mirrored,
// like every other CLI DTO here, so the CLI stays a thin HTTP client.
type crewRunView struct {
	ID             string   `json:"id"`
	SessionID      string   `json:"sessionId"`
	Role           string   `json:"role"`
	Kind           string   `json:"kind"`
	Label          string   `json:"label"`
	Attempt        int      `json:"attempt"`
	Detector       string   `json:"detector"`
	DetectorReason string   `json:"detectorReason"`
	GenAtStart     uint64   `json:"genAtStart"`
	GenAtEnd       uint64   `json:"genAtEnd"`
	StartedAt      string   `json:"startedAt"`
	EndedAt        *string  `json:"endedAt"`
	Outcome        string   `json:"outcome"`
	Result         string   `json:"result"`
	ChangedPaths   []string `json:"changedPaths"`
	HeadSHA        string   `json:"headSha"`
}

type crewRunStartRequest struct {
	Kind  string `json:"kind"`
	Label string `json:"label,omitempty"`
}

type crewRunStartResponse struct {
	Run             crewRunView  `json:"run"`
	Certified       bool         `json:"certified"`
	SupersededRunID string       `json:"supersededRunId"`
	CrewmateRun     *crewRunView `json:"crewmateRun,omitempty"`
}

type crewRunEndRequest struct {
	RunID  string `json:"runId,omitempty"`
	Result string `json:"result,omitempty"`
}

type crewRunEndResponse struct {
	Run         crewRunView `json:"run"`
	Retry       bool        `json:"retry"`
	Attempt     int         `json:"attempt"`
	MaxAttempts int         `json:"maxAttempts"`
	Escalated   bool        `json:"escalated"`
}

type crewRunListResponse struct {
	Runs []crewRunView `json:"runs"`
}

// newCrewRunCommand is the BRACKET a crew member puts around a build, a test
// suite or a device pass.
//
// Two members of a task share one worktree, so a run can read a tree the other
// member is halfway through writing. That is not data loss - it is an
// untrustworthy result, and the dangerous part is that it looks fine. The race
// cannot be prevented (nothing can stop an agent saving a file), so AO detects
// it: a filesystem watcher counts writes to non-ignored paths, `--start` reads
// the counter and `--end` reads it again, and equal readings are what make the
// result trustworthy.
//
// The same bracket is the only way the board can say "qa is running a build"
// rather than the far weaker "qa is awake" - one mechanism, two consumers.
func newCrewRunCommand(ctx *commandContext) *cobra.Command {
	var (
		start  bool
		end    bool
		list   bool
		kind   string
		label  string
		result string
		runID  string
	)
	cmd := &cobra.Command{
		Use:   "run (--start | --end | --list)",
		Short: "Bracket a build or test run so its result can be trusted",
		Long: "Wrap every build, test suite or device pass in a bracket:\n\n" +
			"  ao crew run --start --kind test --label 'go test ./...'\n" +
			"  go test ./...\n" +
			"  ao crew run --end --result pass\n\n" +
			"AO counts writes to the worktree between the two calls. If nothing wrote to a\n" +
			"non-ignored path the result is TRUSTED. If something did, the run is DISCARDED -\n" +
			"which is neither a pass nor a fail: the result read a mixture of two states, so\n" +
			"failing it would blame the code and passing it would launder a mixed result as a\n" +
			"clean one. Discards are re-run automatically, and after three in a row the task\n" +
			"goes to NEEDS YOU for a human to decide.\n\n" +
			"If AO cannot watch the worktree at all it says so and marks the run UNCERTIFIED.\n" +
			"It never quietly falls back to a cheaper check: an absent detector has to be as\n" +
			"visible as one that misses.\n\n" +
			"An open bracket is also how the board and `ao crew status` know this member is\n" +
			"running something rather than merely awake.",
		Example: `  ao crew run --start --kind build          # opens the bracket
  ao crew run --end --result fail          # closes it and records what the build said
  ao crew run --list                       # this session's recent runs`,
		Args: noArgs,
		RunE: func(cmd *cobra.Command, _ []string) error {
			chosen := 0
			for _, on := range []bool{start, end, list} {
				if on {
					chosen++
				}
			}
			if chosen != 1 {
				return usageError{errors.New("choose exactly one of --start, --end or --list")}
			}
			path, err := crewRunPath()
			if err != nil {
				return err
			}
			out := cmd.OutOrStdout()
			switch {
			case start:
				return ctx.startCrewRun(cmd, path, out, kind, label)
			case end:
				return ctx.endCrewRun(cmd, path, out, runID, result)
			default:
				return ctx.listCrewRuns(cmd, path, out)
			}
		},
	}
	cmd.Flags().BoolVar(&start, "start", false, "Open a bracket: read the worktree's write counter and record that a run has begun")
	cmd.Flags().BoolVar(&end, "end", false, "Close the bracket: read the counter again and decide whether the result can be trusted")
	cmd.Flags().BoolVar(&list, "list", false, "List this session's recent bracketed runs")
	cmd.Flags().StringVar(&kind, "kind", "test", "What is about to run: build, test or device")
	cmd.Flags().StringVar(&label, "label", "", "Free-text label for the run, e.g. the command you are about to type")
	cmd.Flags().StringVar(&result, "result", "", "What the build or test said: pass or fail. Omit to record a run that did not judge itself")
	cmd.Flags().StringVar(&runID, "run", "", "Run to close. Defaults to this session's open run")
	return cmd
}

func (c *commandContext) startCrewRun(cmd *cobra.Command, path string, out io.Writer, kind, label string) error {
	var resp crewRunStartResponse
	body := crewRunStartRequest{Kind: strings.TrimSpace(kind), Label: strings.TrimSpace(label)}
	if err := c.postJSON(cmd.Context(), path, body, &resp); err != nil {
		return err
	}
	if resp.SupersededRunID != "" {
		if _, err := fmt.Fprintf(out, "A previous run was never ended; it was closed as uncertified (%s).\n", resp.SupersededRunID); err != nil {
			return err
		}
	}
	if !resp.Certified {
		if _, err := fmt.Fprintf(out,
			"WARNING: no tree-write detector on this worktree - %s\nThis run's result will be marked UNCERTIFIED, not verified.\n",
			resp.Run.DetectorReason); err != nil {
			return err
		}
	}
	// Advisory, and phrased as advice. The other member is not asked to stop and
	// this run is not held: what it is for is the one case nothing here verifies -
	// two `xcodebuild` runs against the SHARED DerivedData that is the whole
	// reason this task has one worktree.
	if peer := resp.CrewmateRun; peer != nil {
		label := peer.Label
		if label == "" {
			label = peer.Kind
		}
		who := peer.Role
		if who == "" {
			who = peer.SessionID
		}
		if _, err := fmt.Fprintf(out,
			"NOTE: %s is already running a %s in this worktree (%s). Nothing stops you sharing it,\nbut two builds against one cache is the case AO does not verify - consider waiting.\n",
			who, peer.Kind, label); err != nil {
			return err
		}
	}
	_, err := fmt.Fprintf(out, "Run %s open (%s). Close it with `ao crew run --end --result pass|fail`.\n", resp.Run.ID, resp.Run.Kind)
	return err
}

func (c *commandContext) endCrewRun(cmd *cobra.Command, path string, out io.Writer, runID, result string) error {
	var resp crewRunEndResponse
	body := crewRunEndRequest{RunID: strings.TrimSpace(runID), Result: strings.TrimSpace(result)}
	if err := c.postJSON(cmd.Context(), path+"/end", body, &resp); err != nil {
		return err
	}
	for _, line := range crewRunEndLines(resp) {
		if _, err := fmt.Fprintln(out, line); err != nil {
			return err
		}
	}
	return nil
}

// crewRunEndLines is what the member is told, and it is deliberately blunt in
// the two cases that matter: a discarded run must never read as the result it
// reported, and an uncertified one must not read as verified.
func crewRunEndLines(resp crewRunEndResponse) []string {
	run := resp.Run
	switch run.Outcome {
	case "discarded":
		lines := []string{
			fmt.Sprintf("DISCARDED - the tree changed under this run, so its result is not trustworthy (attempt %d of %d).", resp.Attempt, resp.MaxAttempts),
		}
		if len(run.ChangedPaths) > 0 {
			lines = append(lines, "  What moved: "+strings.Join(run.ChangedPaths, ", "))
		}
		switch {
		case resp.Escalated:
			lines = append(lines,
				fmt.Sprintf("%d runs discarded in a row. Stop re-running: this task is now NEEDS YOU, and a human decides", resp.MaxAttempts),
				"whether to pause the other member or accept an uncertified result.")
		case resp.Retry:
			lines = append(lines, "Run it again - do not record this result.")
		}
		return lines
	case "uncertified":
		return []string{
			"UNCERTIFIED - nothing was watching the tree, so this result cannot be vouched for.",
			"  Reason: " + run.DetectorReason,
		}
	default:
		verdict := "recorded"
		if run.Result != "" {
			verdict = strings.ToUpper(run.Result)
		}
		return []string{fmt.Sprintf("TRUSTED - the tree did not move; %s stands.", verdict)}
	}
}

func (c *commandContext) listCrewRuns(cmd *cobra.Command, path string, out io.Writer) error {
	var resp crewRunListResponse
	if err := c.getJSON(cmd.Context(), path, &resp); err != nil {
		return err
	}
	if len(resp.Runs) == 0 {
		_, err := fmt.Fprintln(out, "No bracketed runs yet.")
		return err
	}
	for _, run := range resp.Runs {
		if _, err := fmt.Fprintf(out, "%-12s %-7s %-9s %s\n", crewRunStateOf(run), run.Kind, relativeStart(run.StartedAt), run.Label); err != nil {
			return err
		}
	}
	return nil
}

// crewRunStateOf is the CLI's mirror of domain.CrewRun.State: a discarded or
// uncertified run reads as itself, never as the pass/fail it reported.
func crewRunStateOf(run crewRunView) string {
	if run.EndedAt == nil {
		return "running"
	}
	switch run.Outcome {
	case "discarded":
		return "discarded"
	case "uncertified":
		return "uncertified"
	}
	switch run.Result {
	case "pass":
		return "passed"
	case "fail":
		return "failed"
	}
	return "finished"
}

func relativeStart(stamp string) string {
	at, err := time.Parse(time.RFC3339Nano, stamp)
	if err != nil {
		return ""
	}
	return time.Since(at).Round(time.Second).String() + " ago"
}

func crewRunPath() (string, error) {
	sessionID := strings.TrimSpace(os.Getenv("AO_SESSION_ID"))
	if sessionID == "" {
		return "", usageError{errors.New("ao crew run must run inside an AO session (AO_SESSION_ID is not set): a run bracket belongs to the member that made it")}
	}
	return "sessions/" + url.PathEscape(sessionID) + "/crew/runs", nil
}

File: frontend/src/renderer/lib/crew-run.ts
// Pure helpers for the Tests-tab "Machine runs" strip: a crew member's
// bracketed build/test runs and what the tree-write detector concluded about
// each one.
//
// # Why a run has FOUR states and not two
//
// Two members of a task share one worktree, so a build can read a tree the other
// member is halfway through writing. The result is not lost - it is
// untrustworthy, and the dangerous part is that it looks fine. So AO counts
// writes to the worktree across the run, and a run whose tree moved is
// DISCARDED: neither a pass (that is the laundering the whole mechanism exists
// to stop) nor a fail (that would blame the code). A run nothing was watching is
// UNCERTIFIED, for the same reason in the other direction - if a detector that
// misses is bad, an absent one has to be visible too.
//
// The palette follows the tab's rule that the machine's lane is MONOCHROME: the
// pass/fail hues belong to the human's verdict. A discarded or uncertified run
// therefore carries the neutral qa ink, never a verdict colour.

import type { components } from "../../api/schema";
import { tint } from "./comment-inbox";
import { PALETTE as P } from "./smoke-test";

export type CrewRun = components["schemas"]["CrewRun"];

export type CrewRunState = "running" | "passed" | "failed" | "discarded" | "uncertified" | "finished";

/** The one word a run reads as. Mirrors domain.CrewRun.State in the backend. */
export function crewRunState(run: CrewRun): CrewRunState {
	if (!run.endedAt) return "running";
	if (run.outcome === "discarded") return "discarded";
	if (run.outcome === "uncertified") return "uncertified";
	if (run.result === "pass") return "passed";
	if (run.result === "fail") return "failed";
	return "finished";
}

export type CrewRunMeta = {
	/** Chip text. */
	label: string;
	/** One line saying what the state MEANS, so nobody has to infer it. */
	caption: string;
	color: string;
	pillBg: string;
	pillBorder: string;
};

export const CREW_RUN_META: Record<CrewRunState, CrewRunMeta> = {
	running: {
		label: "Running",
		caption: "This member is running something right now.",
		color: P.qaFg,
		pillBg: P.qaBg,
		pillBorder: P.qaBorder,
	},
	passed: {
		label: "Passed",
		caption: "The tree did not move, so what this run saw is what the tree is.",
		color: P.qaFg,
		pillBg: P.qaBg,
		pillBorder: P.qaBorder,
	},
	failed: {
		label: "Failed",
		caption: "The tree did not move, so this failure is real.",
		color: "var(--smoke-fail-fg)",
		pillBg: tint("var(--smoke-fail)", 14),
		pillBorder: tint("var(--smoke-fail)", 40),
	},
	discarded: {
		label: "Discarded",
		caption: "The tree changed under this run, so its result was thrown away - not passed, not failed.",
		color: P.qaFg,
		pillBg: P.qaBg,
		pillBorder: P.qaBorder,
	},
	uncertified: {
		label: "Uncertified",
		caption: "Nothing was watching the tree, so this result cannot be vouched for.",
		color: P.qaFg,
		pillBg: P.qaBg,
		pillBorder: P.qaBorder,
	},
	finished: {
		label: "Finished",
		caption: "The tree did not move. The run recorded no pass or fail of its own.",
		color: P.qaFg,
		pillBg: P.qaBg,
		pillBorder: P.qaBorder,
	},
};

export function crewRunMeta(run: CrewRun): CrewRunMeta {
	return CREW_RUN_META[crewRunState(run)];
}

/** How long the run took, or has been going. Short forms only. */
export function crewRunDuration(run: CrewRun, now: number): string {
	const started = Date.parse(run.startedAt);
	if (Number.isNaN(started)) return "";
	const ended = run.endedAt ? Date.parse(run.endedAt) : now;
	if (Number.isNaN(ended)) return "";
	const secs = Math.max(0, Math.round((ended - started) / 1000));
	if (secs < 60) return `${secs}s`;
	const mins = Math.floor(secs / 60);
	if (mins < 60) return `${mins}m ${secs % 60}s`;
	return `${Math.floor(mins / 60)}h ${mins % 60}m`;
}

/**
 * The CURRENT streak of discarded runs, counted from the newest backwards.
 * Mirrors the daemon's ConsecutiveCrewRunDiscards, which is what the card's lane
 * is derived from, so the strip and the board never disagree.
 *
 * Only a TRUSTED run ends the streak. Two states are SKIPPED rather than
 * counted or treated as a clear:
 *
 * - a run still open, because it has not been judged yet, and reading "not yet
 *   decided" as "fine" would make the escalation vanish the instant the member
 *   started its next attempt - exactly when a person needs to see it;
 * - an UNCERTIFIED run, because nothing watched the tree, so it is not evidence
 *   that this member got a quiet window. Letting it clear the streak would mean
 *   a daemon restart quietly cancelled the escalation.
 */
export function discardStreak(runs: CrewRun[]): number {
	let streak = 0;
	for (const run of runs) {
		const state = crewRunState(run);
		if (state === "running" || state === "uncertified") continue;
		if (state !== "discarded") break;
		streak += 1;
	}
	return streak;
}

/** How many discards are retried automatically before a human has to decide.
 * The backend's domain.CappedRepeat; kept here so the copy can name it. */
export const CREW_RUN_MAX_ATTEMPTS = 3;

/** Whether the streak has spent the automatic retry and parked at NEEDS YOU. */
export function crewRunEscalated(runs: CrewRun[]): boolean {
	return discardStreak(runs) >= CREW_RUN_MAX_ATTEMPTS;
}

/** The label line for one run: its kind, and the command if the member named one. */
export function crewRunTitle(run: CrewRun): string {
	const kind = run.kind ? run.kind[0].toUpperCase() + run.kind.slice(1) : "Run";
	return run.label ? `${kind} · ${run.label}` : kind;
}
